### PR TITLE
feat: add capped future to protos

### LINKED ISSUE #11227: [Capped Futures] Extend the existing protos.
A new CappedFuture message is to be added in the market and proposal payload for the Future Product.

Validation  needs to be added in the commands package.

### LINKED ISSUE #11229: [Capped Futures] Order prices validations
On the following transaction, ensure the price submitted are <= to max_price: 
- order submission
- order amendment
- stop order submission

Any order breaching the max price should be rejected.

On any reprice of pegged orders, any repricing breaching the max_price should be capped to max_price.

### LINKED ISSUE #11230: [Capped Futures] mark price handling
On any mark price updates, whichever mark price configuration is set for the market the new mark price should never exceed the max price configured. This is ensured by filtering out external price data that exceeds the max price, and because orders are rejected if they exceed the price cap, and pegged orders are capped to the max price, so any trade is guaranteed to fall within the `[0, max_price]` range.

### LINKED ISSUE #11231: [Capped Futures] Market settlement
When `binary_settlement` parameter is set to `false` any value `0 <= settlement_price <= max_price` should be accepted as a settlement price.

When `binary_settlement` parameter is set to `true` only `settlement_price=0` or `settlement_price=max_price` should be accepted.

Any other values get ignored and market does not settle, instead it still waits for subsequent values from the settlement oracle until a value which passes the above conditions arrives.

### LINKED ISSUE #11233: [Capped Futures]: ensure capped futures config is included in snapshot state (and correctly restored)
Capped future markets must be loaded from snapshots as capped futures

### PR DESCRIPTION
**Pull request by EVODelavega**
Closes #11227 
Closes #11229 
Closes #11230 
Closes #11231 
Closes #11233

### CODE REVIEW

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -4262,6 +4277,9 @@ enum OrderRejectionReason {
 
   "Pegged orders are not allowed for a party in isolated margin mode"
   ORDER_ERROR_PEGGED_ORDERS_NOT_ALLOWED_IN_ISOLATED_MARGIN_MODE
+
+  "Order price exceeds the max price of a capped future market"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4281]:
```suggestion
  "Order price exceeds the max price of the capped future market"
```

File: protos/sources/vega/vega.proto [hunk 2]
```diff
@@ -534,7 +534,8 @@ enum OrderError {
   ORDER_ERROR_PEGGED_ORDERS_NOT_ALLOWED_IN_ISOLATED_MARGIN_MODE = 52;
   // Order price does not respect market's required tick size
   ORDER_ERROR_PRICE_NOT_IN_TICK_SIZE = 53;
-
+  // Order price exceeds the max price of a capped future market
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:537]:
```suggestion
  // Order price exceeds the max price of the capped future market.
```

**Review by candida-d — DISMISSED** [review]

**Review by witgaw — COMMENTED** [review]
LGTM, just some minor comments

File: core/integration/steps/market/defaults/margin-calculator/default-margin-calculator.json
[the remark this answers is not in the document]
Comment (witgaw) [reply core/integration/steps/market/defaults/margin-calculator/default-margin-calculator.json:1]: Fixed in subsequent commit after I started the review, please ignore.

**Review by witgaw — DISMISSED** [review]

File: commands/proposal_submission.go [hunk 3]
```diff
@@ -1264,6 +1264,25 @@ func checkNewFuture(future *protoTypes.FutureProduct) Errors {
...
+	if future.Cap != nil {
+		// convert to uint to ensure input is valid
+		if len(future.Cap.MaxPrice) == 0 {
+			errs.AddForProperty("new_market.changes.instrument.product.future.cap.max_price", ErrIsRequired)
+		} else {
+			// tick size is validated later on, ignore errors here
+			tick, _ := num.UintFromString(tickSize, 10)
+			if tick.IsZero() {
+				tick = num.UintOne()
+			}
```
Comment (jeremyletang) [inline commands/proposal_submission.go:1276]: couldn't you get a nil pointer here?

Comment (jeremyletang) [reply commands/proposal_submission.go:1276]: eh no actually 0.

File: protos/sources/vega/markets.proto [hunk 4]
```diff
@@ -42,6 +42,17 @@ message Future {
...
+  optional FutureCap cap = 6;
+}
+
+message FutureCap {
+  // Set the maximum price for orders, and settlement data in market decimals.
+  string max_price = 1;
+  // If set to true, the settlement price must either be zero, or equal to the max price.
+  optional bool binary_settlement = 2;
+  // If set to true, positions must be fully collateralised so there is no default risk for any party.
+  optional bool fully_collateralised = 3;
```
Comment (jeremyletang) [inline protos/sources/vega/markets.proto:55]: do these need to be optional? they'd be false by default if not specificied both in JSON or proto? I mean, not sure a optional boolean makes sense by definition.

File: core/execution/future/market.go [hunk 5]
```diff
@@ -3015,7 +3048,13 @@ func (m *Market) resolveClosedOutParties(ctx context.Context, distressedMarginEv
 		// so it'll separate the positions still in distress from the
 		// which have acceptable margins
 		increment := m.tradableInstrument.Instrument.Product.GetMarginIncrease(m.timeService.GetTimeNow().UnixNano())
-		okPos, closed = m.risk.ExpectMargins(distressedMarginEvts, m.lastTradedPrice.Clone(), increment, m.getAuctionPrice())
+		var pr *num.Uint
+		if m.capMax != nil {
+			pr = m.capMax.Clone()
+		} else {
+			pr = m.lastTradedPrice.Clone()
```
[the remark this answers is not in the document]
Comment (EVODelavega) [reply core/execution/future/market.go:3055]: Yes and no, not if the market is fully collateralised, but you're right in that we're not checking for that. Updated that

File: core/execution/common/mark_price.go [hunk 6]
```diff
@@ -222,9 +228,14 @@ func (mpc *CompositePriceCalculator) GetUpdateOraclePriceFunc(oracleIndex int) f
...
 		}
 		p := mpc.scalingFunc(ctx, pd, mpc.oracles[oracleIndex].GetDecimals())
+		// ignore prices that exceed the cap
 		if p == nil || p.IsZero() {
 			return nil
 		}
+		// ignore price data > max price
+		if mpc.maxPrice != nil && p.GT(mpc.maxPrice) {
+			return nil
+		}
```
Comment (jeremyletang) [inline core/execution/common/mark_price.go:238]: Does this need to be ignored, or set to the max price?

File: core/execution/future/market.go [hunk 7]
```diff
@@ -167,6 +167,9 @@ type Market struct {
 	internalCompositePriceCalculator *common.CompositePriceCalculator
 
 	amm *amm.Engine
+
+	fCap   *types.FutureCap
+	capMax *num.Uint
```
Comment (jeremyletang) [inline core/execution/future/market.go:172]: why this one, not use directly fCap.MaxPrice ?

File: commands/proposal_submission.go (hunk 3 above)
Comment (EVODelavega) [reply commands/proposal_submission.go:1276]: No, `UintFromString` never returns nil

File: core/execution/common/mark_price.go (hunk 6 above)
Comment (EVODelavega) [reply core/execution/common/mark_price.go:238]: There's been some debate about this on protocol-design. The conclusion was to ignore values > max price

File: core/execution/future/market.go (hunk 7 above)
Comment (EVODelavega) [reply core/execution/future/market.go:172]: `fCap.MaxPrice` is the value in market precision, `capMax` is in asset precision.

File: protos/sources/vega/markets.proto (hunk 4 above)
Comment (EVODelavega) [reply protos/sources/vega/markets.proto:55]: It doesn't _quite_ make sense, but marked them as optional as they are described as optional in the spec, and in previous PR's there's been comments to mark any fields as optional if the spec calls for it.

File: core/execution/common/mark_price.go (hunk 6 above)
Comment (jeremyletang) [reply core/execution/common/mark_price.go:238]: right.

File: core/execution/future/market.go (hunk 7 above)
Comment (jeremyletang) [reply core/execution/future/market.go:172]: indeed

File: protos/sources/vega/markets.proto (hunk 4 above)
Comment (jeremyletang) [reply protos/sources/vega/markets.proto:55]: right.

**Review by wwestgarth — DISMISSED** [review]

**Review by peterbarrow — DISMISSED** [review]

**Review by candida-d — DISMISSED** [review]

**Review by jeremyletang — DISMISSED** [review]

**Review by jeremyletang — APPROVED** [review]

**Review by candida-d — APPROVED** [review]

**Pull request merged**